File: src/pinverify/verify.ts
import { isSHA, type ActionUpdate } from "../diffparser/parser.js";
import { APIError, type Client } from "../github/client.js";
import { updateOwnerRepo } from "../compare/compare.js";

/** Reports a tag comment that doesn't match the pinned SHA. */
export interface Mismatch {
  update: ActionUpdate;
  tag: string; // the tag from the comment (e.g. "v6.2.0")
  expectSHA: string; // SHA the tag actually resolves to, empty if tag not found
}

/** Sentinel stored in the resolved map when the tag does not exist. */
const TAG_NOT_FOUND = "!not-found";

/**
 * Verifies that new SHA pins match their inline tag comments.
 * Only checks updates where newRef is a SHA and newTag is present.
 */
export async function check(
  client: Client,
  updates: ActionUpdate[]
): Promise<Mismatch[]> {
  interface Job {
    idx: number;
    update: ActionUpdate;
  }

  const jobs: Job[] = [];
  for (let i = 0; i < updates.length; i++) {
    const u = updates[i];
    if (!isSHA(u.newRef) || !u.newTag) continue;
    if (isSHA(u.newTag)) continue;
    jobs.push({ idx: i, update: u });
  }

  if (jobs.length === 0) return [];

  // Deduplicate by action+tag to avoid redundant API calls.
  const resolved = new Map<string, string>();
  const seen = new Set<string>();
  const promises: Promise<void>[] = [];

  for (const j of jobs) {
    const k = `${j.update.action}\0${j.update.newTag}`;
    if (seen.has(k)) continue;
    seen.add(k);

    promises.push(
      (async () => {
        const parsed = updateOwnerRepo(j.update);
        if (!parsed) return;
        const [owner, repo] = parsed;
        try {
          const sha = await client.resolveRefSHA(owner, repo, j.update.newTag);
          resolved.set(k, sha);
        } catch (err) {
          if (err instanceof APIError && (err.statusCode === 404 || err.statusCode === 422)) {
            resolved.set(k, TAG_NOT_FOUND);
          }
          console.warn(
            `warning: could not resolve tag ${j.update.newTag} for ${j.update.action}: ${err}`
          );
        }
      })()
    );
  }

  await Promise.all(promises);

  const mismatches: Mismatch[] = [];
  for (const j of jobs) {
    const k = `${j.update.action}\0${j.update.newTag}`;
    const sha = resolved.get(k);
    if (sha === undefined) continue;
    if (sha === TAG_NOT_FOUND) {
      mismatches.push({
        update: j.update,
        tag: j.update.newTag,
        expectSHA: "",
      });
    } else if (sha !== j.update.newRef) {
      mismatches.push({
        update: j.update,
        tag: j.update.newTag,
        expectSHA: sha,
      });
    }
  }
  return mismatches;
}

/** Returns a human-readable description of a mismatch. */
export function formatMismatch(m: Mismatch): string {
  if (!m.expectSHA) {
    return `${m.update.action}: tag ${m.tag} does not exist, but pinned to ${m.update.newRef.substring(0, 7)}`;
  }
  return `${m.update.action}: tag ${m.tag} resolves to ${m.expectSHA.substring(0, 7)}, but pinned to ${m.update.newRef.substring(0, 7)}`;
}

File: pkg/pinverify/verify.go
// Package pinverify verifies that SHA-pinned action refs match their tag comments.
package pinverify

import (
	"context"
	"fmt"
	"log"
	"strings"
	"sync"

	"github.com/vvoland/gha-pin-diff/pkg/diffparser"
	"github.com/vvoland/gha-pin-diff/pkg/github"
)

// Mismatch reports a tag comment that doesn't match the pinned SHA.
type Mismatch struct {
	Update    diffparser.ActionUpdate
	Tag       string // the tag from the comment (e.g. "v6.2.0")
	ExpectSHA string // SHA the tag actually resolves to
}

// Check verifies that new SHA pins match their inline tag comments.
// Only checks updates where NewRef is a SHA and NewTag is present.
func Check(ctx context.Context, client *github.Client, updates []diffparser.ActionUpdate) []Mismatch {
	type job struct {
		idx    int
		update diffparser.ActionUpdate
	}

	var jobs []job
	for i, u := range updates {
		if !isSHA(u.NewRef) || u.NewTag == "" {
			continue
		}
		if isSHA(u.NewTag) {
			continue
		}
		jobs = append(jobs, job{idx: i, update: u})
	}

	if len(jobs) == 0 {
		return nil
	}

	// Deduplicate by action+tag to avoid redundant API calls.
	type key struct{ action, tag string }
	resolved := make(map[key]string)
	var mu sync.Mutex
	var wg sync.WaitGroup

	seen := make(map[key]bool)
	for _, j := range jobs {
		k := key{j.update.Action, j.update.NewTag}
		if seen[k] {
			continue
		}
		seen[k] = true
		wg.Go(func() {
			owner, repo, ok := actionOwnerRepo(j.update.Action)
			if !ok {
				return
			}
			sha, err := client.ResolveRefSHA(ctx, owner, repo, j.update.NewTag)
			if err != nil {
				log.Printf("warning: could not resolve tag %s for %s: %v", j.update.NewTag, j.update.Action, err)
				return
			}
			mu.Lock()
			resolved[k] = sha
			mu.Unlock()
		})
	}
	wg.Wait()

	var mismatches []Mismatch
	for _, j := range jobs {
		k := key{j.update.Action, j.update.NewTag}
		sha, ok := resolved[k]
		if !ok {
			continue
		}
		if sha != j.update.NewRef {
			mismatches = append(mismatches, Mismatch{
				Update:    j.update,
				Tag:       j.update.NewTag,
				ExpectSHA: sha,
			})
		}
	}
	return mismatches
}

func isSHA(s string) bool {
	if len(s) != 40 {
		return false
	}
	for _, c := range s {
		if (c < '0' || c > '9') && (c < 'a' || c > 'f') {
			return false
		}
	}
	return true
}

func actionOwnerRepo(action string) (string, string, bool) {
	owner, rest, ok := strings.Cut(action, "/")
	if !ok {
		return "", "", false
	}
	repo, _, _ := strings.Cut(rest, "/")
	return owner, repo, true
}

// FormatMismatch returns a human-readable description of a mismatch.
func FormatMismatch(m Mismatch) string {
	return fmt.Sprintf("%s: tag %s resolves to %s, but pinned to %s",
		m.Update.Action, m.Tag, m.ExpectSHA[:7], m.Update.NewRef[:7])
}
